### ISSUE TITLE
Add link context to structure menu link search

### ISSUE THREAD

**Issue by bewoldt** [uiowa/uiowa#4235]
# Story

As an editor, I would like to be able to link content on block types as well as the menu management interface using the autocomplete functionality found in the WYSIWYG.

# Problem
Within a Structure > menu page, when you use the search to choose where to link the page to, it does not provide context on whether it is a page, article, or person.

<img width="666" alt="Screen Shot 2021-09-30 at 1 21 33 PM" src="https://user-images.githubusercontent.com/731952/135514008-4e8142ad-9eac-49a1-85f3-89d37a5c8a4e.png">

# Steps to reproduce
Go into  Structure > Menus and edit/add a page and use the search menu to choose a link destination.

**Comment by joewhitsitt** [comment]
Capturing what I responded with in email:

> Unfortunately, not all link fields are the same. This might be possible, but I would have to dig a bit.
> A workaround as it relates to the menus is to use the node edit form to create/manage the link.

This is Drupal core default.

**Comment by joewhitsitt** [comment]
Long time issue to bring linkit to the link field. https://www.drupal.org/project/linkit/issues/2712951

**Comment by briand44** [comment]
We will want to provide this context to other component link fields as well (cards, call to action, etc.)

**Comment by joewhitsitt** [comment]
This took a big step forward today https://www.drupal.org/project/linkit/issues/2712951#comment-14961986

Many have said they have been using this in production for years, but a stable release is on the horizon.

**Comment by pyrello** [comment]
After #6539 is merged, this is un-blocked.

**Comment by pyrello** [comment]
> After #6539 is merged, this is un-blocked.

Actually, I was too quick responding to this. I'm not sure if this would apply to the menu system or not.

**Comment by joewhitsitt** [comment]
yeah, menu edit forms are not configurable through the UI. i think this could be done in a form_alter though similar to https://github.com/uiowa/uiowa/blob/main/docroot/modules/custom/uiowa_core/src/Plugin/Block/SearchBlock.php#L91

We already tap in to the form here `sitenow_form_menu_link_content_form_alter()`

<img width="694" alt="Screenshot 2023-06-13 at 4 53 13 PM" src="https://github.com/uiowa/uiowa/assets/4663676/c2495e8c-bac3-43d8-80d6-c1d528ff1462">

**Comment by briand44** [comment]
Grooming conversation.... Menu system, card, call to action, other default block types. Possibly add in linking to unpublished if time allows. Sized as large pending Joe scanning for which block types should be in scope here.

**Comment by joewhitsitt** [comment]
To provide more clarity in sizing during planning, in addition to the menu form, I _think_ these are the ones related to _default_ block types:

field_collection_more_path
field_uiowa_banner_link
field_uiowa_button_link
field_uiowa_card_link
field_uiowa_cta_link
field_uiowa_event_link
field_uiowa_image_link
field_uiowa_quote_citation
field_timeline_link
field_uiowa_slide_link

**Comment by joewhitsitt** [comment]
@bspeare discovered that `<nolink>` is not working with this. even with the patch (https://www.drupal.org/project/linkit/issues/3223781) and new matchers configured, the menu link form alter have in place does not work. Additionally, I worry there might need to be an update hook with this as linkit modifies the value stored in the db. `<front>` to `\` and `<nolink>` to `route:<nolink>`

Created a rollback PR https://github.com/uiowa/uiowa/pull/6763 until we can revisit this as planned work.

**Comment by briand44** [comment]
I had a couple requests semi related to this today. Registrar asked for the ability to have additional context when using the featured content block as they are running into contacts that have the same name as other content in their site.

Also, not sure if this was part of the previous work here but Tippie is asking for the ability to link to unpublished entities from the card component, similarly to how they can do that through the WYSIWYG in the text area component.

**Comment by bspeare** [comment]
This pr further documents the issues with linkit and the card component: https://github.com/uiowa/uiowa/pull/7529.

**Comment by e-marie-w** [comment]
Grooming Convo: still need to fix issues with linkit and cards. Check issue queue, maybe it's been fixed upstream? Another option is to leave the card out at this time, or change how we structure cards. PR https://github.com/uiowa/uiowa/pull/7529 in previous comment has linkit working for cards, but is handled through additional processing in Card.php. Linkit is being added to core but probably not for a long while yet.

**Comment by briand44** [comment]
Sounds like there is a medium amount of work left.

**Comment by joewhitsitt** [comment]
Questions for the team:

- Are the default list blocks (articles, people, pages) in scope or just the blocks visible in the Add Component popup?
- Is it okay to move forward with partial LinkIt support on menu links? Absolute URLs to media files will still work, but there won't be autocomplete matching and conversion of /media/file/123 type stuff.

**Comment by joewhitsitt** [comment]
Talked with Brian a bit about https://github.com/uiowa/uiowa/issues/4235#issuecomment-2565677952

- Brian thinks we can poke a bit to see if the list blocks are an easy add to this work, but if not, slicing it out would make sense.
- Brian also thinks that moving forward with any context for menu links even if media is absent is a good step forward.

Slight concern about our "upgrade" path if/when the Linkit functionality moves into core. Keeping our customizations here limited will reduce the number of problems to solve later.

**Comment by joewhitsitt** [comment]
I did some quick poking at this and realized that adding it to the list blocks runs into the same issue of having to create our own rendering logic to handle external/internal/media paths. I also realized we ~~are already doing some of this work~~ established a foundation for this stuff (docroot/modules/custom/uiowa_core/src/LinkHelper.php).

So maybe we can combine/append/consolidate what we are doing here, the card, the timeline block, listblocks and the media helper function Ben had originally put in https://github.com/uiowa/uiowa/pull/7529/commits/46fe834001ec89954a3d53ba409f74049b6e176f

Update: This got out of control as the Field API and Form API versions don't need the same things. Realized I was creating too much unnecessary bloat.
